File: merge_sort.py
import random


def merge_sort(array:list) -> None:
    array_length = len(array)

    if(array_length < 2):
        return
    
    mid_point = array_length // 2

    left_half = array[:mid_point]
    right_half = array[mid_point:]

    merge_sort(left_half)
    merge_sort(right_half)
    
    left_size = len(left_half)
    right_size = len(right_half)
    i,j,k = 0,0,0

    while(i < left_size  and j < right_size):
        if(left_half[i] < right_half[j]):
            array[k] = left_half[i]
            i += 1
        else:
            array[k] = right_half[j]
            j += 1

        k += 1
    
    while(i < left_size):
        array[k] = left_half[i]
        i += 1
        k += 1
    
    while (j < left_size):
        array[k] = right_half[j]
        j += 1
        k += 1


    

array = [random.randint(0,1000000) for _ in range(10000000)]




merge_sort(array)
        

File: merge_sort/MergeSort.java
package merge_sort;

import java.util.Arrays;
import java.util.Random;

public class MergeSort {

    public static void main(String[] args) {

        Random random = new Random();
        int[] numbers = new int[100];

        for (int i = 0; i < numbers.length; i++) {
            numbers[i] = random.nextInt(100);
        }

        System.out.println(Arrays.toString(numbers));
        mergeSort(numbers);
        System.out.println(Arrays.toString(numbers));
    }

    public static void mergeSort(int[] array) {
        int arrayLength = array.length;
        if (arrayLength < 2) {
            return;
        }
        int midIndex = arrayLength / 2;

        int[] leftHalf = new int[midIndex];
        int[] rightHalf = new int[arrayLength - midIndex];

        for (int i = 0; i < midIndex; i++) {
            leftHalf[i] = array[i];
        }

        for (int i = midIndex; i < arrayLength; i++) {
            rightHalf[i - midIndex] = array[i];
        }

        mergeSort(leftHalf);
        mergeSort(rightHalf);

        merge(array, leftHalf, rightHalf);
    }

    private static void merge(int[] array, int[] leftHalf, int[] rightHalf) {

        int leftSize = leftHalf.length;
        int rightSize = rightHalf.length;
        int i = 0, j = 0, k = 0;

        while (i < leftSize && j < rightSize) {
            if (leftHalf[i] < rightHalf[j]) {
                array[k] = leftHalf[i];
                i++;
            } else {
                array[k] = rightHalf[j];
                j++;
            }

            k++;
        }

        while (i < leftSize) {
            array[k] = leftHalf[i];
            i++;
            k++;

        }
        while (j < rightSize) {
            array[k] = rightHalf[j];
            j++;
            k++;
        }
    }

}